HiP-DataStore/Controllers/TagsController.cs

```diff
@@ -0,0 +1,197 @@
+        [ProducesResponseType(400)]
+        [ProducesResponseType(422)]
+        [ProducesResponseType(409)]
+        public async Task<IActionResult> PostAsync([FromBody]TagArgs tag)
+        {
+            if (!ModelState.IsValid)
+                return BadRequest();
+
+            if (tag.Image != null&& !_mediaIndex.IsPublishedImage(tag.Image.Value)) 
+                return StatusCode(422, ErrorMessages.ImageNotFoundOrNotPublished(tag.Image.Value));
+
+            int id = _entityIndex.NextId(ResourceType.Tag);
+            var ev = new TagCreated()
+            {
+                Id = id,
+                Properties = tag
+            };
+
+            await _ev.AppendEventAsync(ev);
+
+            if (tag.Image != null)
+            {
+                var newRef = new ReferenceAdded(ResourceType.Tag, id, ResourceType.Media, tag.Image.Value);
+                await _ev.AppendEventAsync(newRef);
+            }
+
+            return Ok(id);
+
+        }
+
+        [HttpGet]
+        [ProducesResponseType(typeof(AllItemsResult<TagResult>), 200)]
+        [ProducesResponseType(304)]
+        [ProducesResponseType(422)]
+        public IActionResult GetAll(TagQueryArgs args)
+        {
+            if (!ModelState.IsValid)
+                return BadRequest(ModelState);
+
+            var query = _db.Database.GetCollection<Tag>(ResourceType.Tag.Name).AsQueryable();
+
+            try
+            {
+                // TODO Add filtering by timestamp
+                var tags = query
+                    .FilterByIds(args.ExcludedIds, args.IncludedIds)
+                    .FilterByStatus(args.Status)
+                    .FilterByUsage(args.Used)
+                    .FilterIf(!string.IsNullOrEmpty(args.Query), x =>
+                        x.Title.ToLower().Contains(args.Query.ToLower()) ||
+                        x.Description.ToLower().Contains(args.Query.ToLower()))
+                    .Sort(args.OrderBy,
+                        ("id", x => x.Id),
+                        ("title", x => x.Title),
+                        ("timestamp", x => x.Timestamp))
+                    .PaginateAndSelect(args.Page, args.PageSize, x => TagResult.ConvertFromTag(x));
+                   
+        
+                return Ok(tags);
+            }
+            catch (InvalidSortKeyException e)
+            {
+                return StatusCode(422, e.Message);
+            }
+        }
+
+        [HttpGet("{id}")]
+        [ProducesResponseType(typeof(TagResult),200)]
+        [ProducesResponseType(304)]
+        [ProducesResponseType(400)]
+        [ProducesResponseType(404)]
+        public IActionResult GetById(int id,DateTimeOffset? timestamp = null)
+        {
+            if (!ModelState.IsValid)
+                return BadRequest();
+
+            var query = _db.Database.GetCollection<Tag>(ResourceType.Tag.Name).AsQueryable();
+
+            var tag = query.Where(x => x.Id == id).FirstOrDefault();
```

Comment:
instead of ".Where(condition).FirstOrDefault()" you can use ".FirstOrDefault(condition)"